File: sim/motion.py
"""Scripted base motion: sinusoidal torso disturbance, world frame.

The torso mocap body is driven as home + A*sin(2*pi*f*t) in position
(world xyz) and orientation (rpy). The EE targets were resolved to the
world frame once at startup, so this motion IS the disturbance — the
task stays a world-frame pose hold.

Mocap writes teleport the torso: mocap bodies carry no velocity state,
so the arms feel base motion as position steps. Accepted phase-1 gap.
"""

import mujoco
import numpy as np

from controller.transforms import (
    angular_velocity_from_rpy_rates,
    rotation_from_rpy,
    sine_offset,
    sine_rate,
)
from sim import world

# --- research levers ---------------------------------------------------
# Edit these SI values to change the default base-motion scenario used by
# the main viewer and general-purpose analysis views.
LINEAR_AMPLITUDE = np.array([0.0, 0.00, 0.0])
ROTATIONAL_AMPLITUDE = np.array([0.0, 0.0, 0.0])
LINEAR_FREQUENCY = 0.5
ROTATIONAL_FREQUENCY = 0.5

_TORSO_MOCAP_IDX = world.model.body_mocapid[world.torso_body_id]
HOME_POS = world.data.mocap_pos[_TORSO_MOCAP_IDX].copy()
# torso_pose_at's elementwise rpy sum is only exact because the home
# rotation is identity (cf. target_motion.py's matrix composition); a
# tilted torso home in scene.xml must fail here, not silently bend the
# scripted motion.
assert np.allclose(world.data.mocap_quat[_TORSO_MOCAP_IDX],
                   [1.0, 0.0, 0.0, 0.0]), \
    "torso home rotation is not identity; HOME_RPY = zeros is invalid"
HOME_RPY = np.zeros(3)


def torso_pose_at(t, linear_amplitude=LINEAR_AMPLITUDE,
                  linear_frequency=LINEAR_FREQUENCY,
                  rotational_amplitude=ROTATIONAL_AMPLITUDE,
                  rotational_frequency=ROTATIONAL_FREQUENCY):
    """(pos, rpy) of the scripted torso at time t — pure math."""
    pos = HOME_POS + sine_offset(t, linear_amplitude, linear_frequency)
    rpy = HOME_RPY + sine_offset(t, rotational_amplitude,
                                 rotational_frequency)
    return pos, rpy


def torso_twist_at(t, linear_amplitude=LINEAR_AMPLITUDE,
                   linear_frequency=LINEAR_FREQUENCY,
                   rotational_amplitude=ROTATIONAL_AMPLITUDE,
                   rotational_frequency=ROTATIONAL_FREQUENCY):
    """(v (3,) m/s, w (3,) rad/s) of the scripted torso at time t, world
    frame — the analytic time derivative of torso_pose_at. This is the
    base twist the mocap teleports never give the simulator; on hardware
    the same quantity comes from Vicon."""
    v = sine_rate(t, linear_amplitude, linear_frequency)
    rpy = HOME_RPY + sine_offset(t, rotational_amplitude,
                                 rotational_frequency)
    rpy_dot = sine_rate(t, rotational_amplitude, rotational_frequency)
    return v, angular_velocity_from_rpy_rates(rpy, rpy_dot)


def set_torso_pose(t, linear_amplitude=LINEAR_AMPLITUDE,
                   linear_frequency=LINEAR_FREQUENCY,
                   rotational_amplitude=ROTATIONAL_AMPLITUDE,
                   rotational_frequency=ROTATIONAL_FREQUENCY):
    """Write the scripted torso pose into the mocap body.

    All-zero amplitudes: no write at all — the static condition, and the
    torso stays free (hand-draggable in the viewer as a perturbation)."""
    if not (np.any(linear_amplitude) or np.any(rotational_amplitude)):
        return
    pos, rpy = torso_pose_at(t, linear_amplitude, linear_frequency,
                             rotational_amplitude, rotational_frequency)
    quat = np.zeros(4)
    mujoco.mju_mat2Quat(quat, rotation_from_rpy(rpy).flatten())
    world.data.mocap_pos[_TORSO_MOCAP_IDX] = pos
    world.data.mocap_quat[_TORSO_MOCAP_IDX] = quat

File: cpp/src/sim/Motion.cpp
#include "sim/Motion.h"

#include <stdexcept>

#include <mujoco/mujoco.h>

#include "math/Transforms.h"

namespace srl::sim {

std::pair<Eigen::Vector3d, Eigen::Vector3d> TorsoMotion::PoseAt(double t) const {
  // The element-wise rpy sum is only exact because the home rotation is
  // identity (contrast TargetMotion's matrix composition); CaptureTorsoHome
  // enforces that.
  const Eigen::Vector3d position =
      home_position +
      transforms::SineOffset(t, linear_amplitude, linear_frequency);
  const Eigen::Vector3d rpy =
      transforms::SineOffset(t, rotational_amplitude, rotational_frequency);
  return {position, rpy};
}

std::pair<Eigen::Vector3d, Eigen::Vector3d> TorsoMotion::TwistAt(double t) const {
  const Eigen::Vector3d linear =
      transforms::SineRate(t, linear_amplitude, linear_frequency);
  const Eigen::Vector3d rpy =
      transforms::SineOffset(t, rotational_amplitude, rotational_frequency);
  const Eigen::Vector3d rpy_dot =
      transforms::SineRate(t, rotational_amplitude, rotational_frequency);
  return {linear, transforms::AngularVelocityFromRpyRates(rpy, rpy_dot)};
}

TorsoMotion CaptureTorsoHome(const MujocoBackend& backend) {
  const mjModel* model = backend.model();
  const mjData* data = backend.data();
  const int torso_body = mj_name2id(model, mjOBJ_BODY, "torso");
  if (torso_body < 0) throw std::runtime_error("'torso' not in scene");
  const int mocap = model->body_mocapid[torso_body];

  TorsoMotion motion;
  motion.home_position = Eigen::Vector3d(data->mocap_pos[3 * mocap + 0],
                                         data->mocap_pos[3 * mocap + 1],
                                         data->mocap_pos[3 * mocap + 2]);
  const Eigen::Vector4d home_quat(
      data->mocap_quat[4 * mocap + 0], data->mocap_quat[4 * mocap + 1],
      data->mocap_quat[4 * mocap + 2], data->mocap_quat[4 * mocap + 3]);
  if ((home_quat - Eigen::Vector4d(1.0, 0.0, 0.0, 0.0)).cwiseAbs().maxCoeff() >
      1e-8) {
    throw std::runtime_error(
        "torso home rotation is not identity; HOME_RPY = zeros is invalid");
  }
  return motion;
}

void InstallTorsoDriver(MujocoBackend& backend, TorsoMotion motion) {
  backend.ConfigureTorsoDriver(
      [motion](double t) { return motion.PoseAt(t); },
      [motion](double t) { return motion.TwistAt(t); });
}

}  // namespace srl::sim
